WIP: Add a TOC for OpenJ9 compiler documents

a7ehuo opened the pull request:
Add a TOC for OpenJ9 compiler documents

Signed-off-by: Annabelle Huo <Annabelle.Huo@ibm.com>

a7ehuo commented:
@vijaysun-omr @0xdaryl I have created this TOC to organize our documents more systematically, hoping it would also help us identify topics to be documented in the future. It should be a living document and updated when new documents are added under [openj9/doc/compiler](https://github.com/eclipse-openj9/openj9/tree/master/doc/compiler) and [omr/doc/compiler](https://github.com/eclipse/omr/tree/master/doc/compiler). May I ask you to review? Thank you! 

The TOC contains:
- Links to all the compiler documents under [openj9/doc/compiler](https://github.com/eclipse-openj9/openj9/tree/master/doc/compiler) and [omr/doc/compiler](https://github.com/eclipse/omr/tree/master/doc/compiler) 
- Links to some of the lightening talks and OpenJ9 blog articles in debug and memory, etc.
- Links to some of the open issues on documentation topics such as [DLT](https://github.com/eclipse-openj9/openj9/issues/12505), etc.
- Non-hyperlinks are the items I've identified would be nice to be documented. They should be reviewed to see if they are important or relevant enough to require documentation

a7ehuo commented:
@dsouzai @gita-omr I’m wondering if I could also get some feedback/input from you on this PR. Thank you in advance!

vijaysun-omr commented:
Thanks for taking the initiative to organize our documentation Annabelle. 

Some thoughts come to mind when looking at the topics:

1. I feel there are some other concepts such as CFG, blocks, structures etc. that could be covered in the initial section that talks about IL basics such as symbol references, nodes etc. I felt that "extending trees" was an unintuitive name for a section that is really more about what an IL opcode is and some guidelines on when to add a new IL opcode.

2. You may want to annotate topics that can belong in OMR from those ones that need to belong in OpenJ9. The next issue from that will be to make sure that a unified view is easy for a reader to form somehow (maybe via a handy link from OpenJ9 to documentation that lives in OMR, but in the right context of the documentation from an OpenJ9 perspective, so that people don't have to search too hard)

3. You may need to think through what some of the basic structure of a) each article and b) each section should be, i.e. think about the goal for each article/section. Your effort should lead to less ad-hoc-ness in terms of the topics we are choosing to document and how/where they are organized, but there may still be ad-hoc-ness in what the article contents are, and we want to have some basic aspects get covered in every article (rather than have a vast disparity on the quality of the material).

dsouzai commented:
Some initial thoughts I have:

1. I would put Memory into its own section because there's enough information there that it probably shouldn't be hidden away in the misc section.
2. Nit pick, but it might read better to reorder the topics, something like moving "Compilation Control" after "Code Generator", and "AOT"/"JITServer" after "Debug".

Other than that, it looks like a good starting point.

0xdaryl commented:
Thanks for getting this started!  I agree with Vijay's point about separating OMR and OpenJ9 optimizations.  Also, separating the analysis passes versus the optimizations might be useful.

There probably be an overall compiler architecture section that describes everything at a high level and how the pieces fit and flow together:  JIT Control, ILGen, Optimizer, CodeGen, Runtime (AOT, Profiler, JIT runtime for dispatches and data resolution, recompilation, ...)

a7ehuo commented:
Addressing some of the [comments](https://github.com/eclipse-openj9/openj9/pull/14572#issuecomment-1048827591) from @vijaysun-omr 

> 1. ... there are some other concepts such as CFG, blocks, structures etc. that could be covered in the initial section that talks about IL basics such as symbol references, nodes etc.

Some of these concepts are covered in [Testarossa's Intermediate Language: An Intro to Trees (OMR)](https://github.com/eclipse/omr/blob/master/doc/compiler/il/IntroToTrees.md). I think either that document could be updated to be reorganized in a better way to give an introduction on nodes, trees, blocks, CFGs, and structures. Or we could create a new doc to cover the introduction on these concepts. 

> I felt that "extending trees" was an unintuitive name for a section that is really more about what an IL opcode is and some guidelines on when to add a new IL opcode.''

Good point! I'll update the link name to [ExtendingTrees.md](https://github.com/eclipse/omr/blob/master/doc/compiler/il/ExtendingTrees.md) `Things to Consider When Adding a New IL Opcode`. 

> 2. You may want to annotate topics that can belong in OMR from those ones that need to belong in OpenJ9 ... 

I'm hoping this TOC that resides in OpenJ9 gives a unified view for all compiler documents from OMR and OpenJ9. I plan to annotate each topic with `(OMR)` or `(OpenJ9)` like below. 
```
• Overview of IL Generator (OMR)
• Testarossa's Intermediate Language: An Intro to Trees (OMR)
• Node Transmutation (OMR)
• …
• IL FAQs (OpenJ9)
```

> 3. ... less ad-hoc-ness in terms of the topics we are choosing to document and how/where they are organized ...

Most of the undocumented topics under 2. Optimizer for now are kind of ad-hoc. I'll clean them up since the first version of TOC should really focusing on reflecting what and where our existing documents are, plus some future documents that cover basic key concepts.

a7ehuo commented:
>"AOT"/"JITServer" after (...or maybe even before) "Debug".

@dsouzai Sorry I'm not sure I follow the [comment](https://github.com/eclipse-openj9/openj9/pull/14572#issuecomment-1048948539). AOT(6) and JITServer(7) are currently arranged before Debug (9). The logic is to introduce all the code components before debugging.

dsouzai commented:
> The logic is to introduce all the code components before debugging.

Yeah that's fair; I was thinking more I guess something like, Compilation Control after Compiler components, then Profiling/Runtime, then AOT/JITServer, and then Debug/Misc.

a7ehuo commented:
> ... separating the analysis passes versus the optimizations might be useful.

Regarding to the above [comment](https://github.com/eclipse-openj9/openj9/pull/14572#issuecomment-1049062337), @0xdaryl I'm not sure I follow what is meant by separating the analysis passes from optimizations. Could you elaborate? 

> There probably be an overall compiler architecture section that describes everything at a high level and how the pieces fit and flow together: JIT Control, ILGen, Optimizer, CodeGen, Runtime (AOT, Profiler, JIT runtime for dispatches and data resolution, recompilation, ...)

It can be done in two ways:
(1) There is already a document that is listed as `Overview of OpenJ9 JIT Compiler` as the first item on the TOC. These topics could be cover in that document seperately. 
Or/And
(2) Add a brief description in this TOC to give a high level description of each top level code component and how they fit together

vijaysun-omr commented:
Just wanted to re-emphasize that my main concern in comment 3 was "there may still be ad-hoc-ness in what the article contents are, and we want to have some basic aspects get covered in every article (rather than have a vast disparity on the quality of the material).".

0xdaryl commented:
> Separating the analysis passes versus the optimizations might be useful.

The optimizer deals with high-level optimizations, but also analyses that enable those optimizations.  For example, forward/backward data flow analyses, alias analysis, value numbering, etc.  I thought it would be useful to separate those from the actual optimizations.  At one point we actually had them in a separate "analyses" subdirectory under the optimizer.

gita-omr commented:
Hi @a7ehuo, this is something absolutely incredible! Thanks so much for organizing and putting it all together. I like the structure. Just one idea:  when I learn something I first want to know the main data structures. I would add something like "Fundamental Data Structrures" at the very beginning. This would include:

- openj9 object model
- C vs Java stack 
- some others ?

These do not belong to JIT but something JIT operates on. We can put links to the VM documents that describe those. Inside the individual JIT documents, I would also suggest that pattern: (1) data structures, (2) algorithms.

I might think of something else but these are high level comments from me. Thanks!

a7ehuo commented:
Thanks for all the suggestion @gita-omr! I've added a section "Fundamental Data Structures JIT Operates on" as the first chapter of the TOC. 

@vijaysun-omr @0xdaryl I've addressed all the comments and added more links to some of the training videos on YouTube. It's ready for another review. Thank you!

dsouzai commented on doc/compiler/README.md:
```diff
@@ -0,0 +1,211 @@
+levels based on its hotness such as cold, warm, hot, scorching etc.
+
+Code generators translate IL trees to assembly instructions for the target architecture.
+They first generate machine instructions that will cause trees' values to appear in virtual registers.
```
```suggestion
They first generate machine instructions wherein the result of each (value-producing) IL node is placed in a virtual register.
```

a7ehuo replied:
Addressed in d6c45e0

dsouzai reviewed (changes requested):
Some minor changes requested.

a7ehuo commented on doc/compiler/README.md:
```diff
@@ -0,0 +1,211 @@
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+## Table of Content
```
Addressed in d6c45e0

a7ehuo commented on doc/compiler/README.md:
```diff
@@ -0,0 +1,211 @@
+## Table of Content
+---
+
+When a method is chosen to be compiled, it is handed to IL(Intermediate Language) generator first.
```
Addressed in d6c45e0

a7ehuo commented on doc/compiler/README.md:
```diff
@@ -0,0 +1,211 @@
+---
+
+When a method is chosen to be compiled, it is handed to IL(Intermediate Language) generator first.
+IL generator translates Java bytecodes to IL trees.
```
Addressed in d6c45e0

a7ehuo commented on doc/compiler/README.md:
```diff
@@ -0,0 +1,211 @@
+Next, IL trees are processed by many independent optimization passes in the Optimizer.
+The optimizations first analyze the IL trees to find opportunities to optimize. Secondly,
+they transform the IL trees for better performance. Each method is compiled at different optimization
+levels based on its hotness such as cold, warm, hot, scorching etc.
```
Addressed in d6c45e0

dsouzai commented:
@vijaysun-omr @0xdaryl I'd like to merge this PR to give us a starting point from which it can be added to. Are there any outstanding concerns you have for the PR as it stands?

vijaysun-omr commented:
I have approved this so that it can be merged since many of our suggestions are now incorporated. Thanks again @a7ehuo for your excellent initiative.

0xdaryl commented on doc/compiler/README.md:
```diff
@@ -0,0 +1,211 @@
+                    |  +-------------------v------------------+ |
+                    |  |            Code Generators           | |
+                    |  | +-----+ +---+ +---------+ +--------+ | |
+                    |  | | X86 | | Z | | PowerPC | | Aarch  | | |
```
I would say "ARM" rather than "Aarch".

a7ehuo replied:
Addressed in e8a650d

0xdaryl commented on doc/compiler/README.md:
```diff
@@ -0,0 +1,211 @@
+                    |  +-------------------+------------------+ |
+                    |                      |                    |
+                    +----------------------+--------------------+
++-----------------------------+      +-----+-----+
```
Is there a connecting line missing from the JIT to the Metadata+Code ?

a7ehuo replied:
Addressed in e8a650d

0xdaryl commented on doc/compiler/README.md:
```diff
@@ -0,0 +1,211 @@
+    * [X86 Binary Encoding Scheme (OMR)](https://github.com/eclipse/omr/blob/master/doc/compiler/x/OpCodeEncoding.md)
+  * PowerPC
+  * s390
+  * Aarch
```
Aarch -> ARM

a7ehuo replied:
Addressed in e8a650d

0xdaryl reviewed (approved):
Great start!  A couple of minor nits from me.  Otherwise I approve.

a7ehuo commented:
@0xdaryl @dsouzai All comments are addressed. Ready for another review. Thanks!
